plugins/modem-manager/fu-plugin-modem-manager.c

```diff
@@ -0,0 +1,472 @@
+	}
+	/* create device and add to cache */
+	dev = fu_mm_device_udev_new (priv->manager,
+				     priv->inhibited->physical_id,
+				     priv->inhibited->vendor,
+				     priv->inhibited->name,
+				     priv->inhibited->version,
+				     (const gchar **) priv->inhibited->guids,
+				     priv->inhibited->update_methods,
+				     priv->inhibited->detach_fastboot_at,
+				     priv->inhibited->port_at_ifnum);
+	fu_mm_device_udev_add_port (dev, subsystem, path, ifnum);
+	fu_plugin_cache_add (plugin, priv->inhibited->physical_id, dev);
+
+	/* wait a bit before probing, in case more ports get added */
+	fu_plugin_mm_udev_device_ports_timeout_reset (plugin);
+}
+
+static gboolean
+fu_plugin_mm_udev_uevent_cb (GUdevClient	*udev,
+			     const gchar	*action,
+			     GUdevDevice	*device,
+			     gpointer		 user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN (user_data);
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	const gchar *subsystem = g_udev_device_get_subsystem (device);
+	const gchar *name = g_udev_device_get_name (device);
+	g_autofree gchar *path = NULL;
+	g_autofree gchar *device_sysfs_path = NULL;
+	gint ifnum = -1;
+
+	if ((action == NULL) || (subsystem == NULL) || (priv->inhibited == NULL) || (name == NULL))
+		return TRUE;
+
+	/* ignore if loading port info fails */
+	if (!fu_mm_utils_get_udev_port_info (device, &device_sysfs_path, &ifnum, NULL))
+		return TRUE;
+
+	/* ignore all events for ports not owned by our device */
+	if (g_strcmp0 (device_sysfs_path, priv->inhibited->physical_id) != 0)
+		return TRUE;
+
+	/* ignore non-cdc-wdm usbmisc ports */
+	if (g_str_equal (subsystem, "usbmisc") && !g_str_has_prefix (name, "cdc-wdm"))
+		return TRUE;
+
+	path = g_strdup_printf ("/dev/%s", name);
+
+	if ((g_str_equal (action, "add")) || (g_str_equal (action, "change"))) {
+		g_debug ("added port to inhibited modem: %s (ifnum %d)", path, ifnum);
+		fu_plugin_mm_udev_device_port_added (plugin, subsystem, path, ifnum);
+	} else if (g_str_equal (action, "remove")) {
+		g_debug ("removed port from inhibited modem: %s", path);
+		fu_plugin_mm_udev_device_removed (plugin);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_mm_inhibit_device (FuPlugin *plugin, FuDevice *device, GError **error)
+{
+	static const gchar *subsystems[] = { "tty", "usbmisc", NULL };
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	g_autofree gchar *guids_str = NULL;
+	g_autoptr(FuPluginMmInhibitedDevice) info = NULL;
+	const gchar *inhibition_uid;
+
+	fu_plugin_mm_uninhibit_device (plugin);
+
+	info = g_new0 (FuPluginMmInhibitedDevice, 1);
+	info->physical_id = g_strdup (fu_device_get_physical_id (device));
+	info->vendor = g_strdup (fu_device_get_vendor (device));
+	info->name = g_strdup (fu_device_get_name (device));
+	info->version = g_strdup (fu_device_get_version (device));
+	guids_str = fu_device_get_guids_as_str (device);
+	info->guids = g_strsplit (guids_str, ",", -1);
+	info->update_methods = fu_mm_device_get_update_methods (FU_MM_DEVICE (device));
```

Comment:
This seems like a hack? Can't we just add the guids as a gchar** or even just ref the GPtrArray?

Reply:
yes, let me look at that.

Reply:
done